man/rmdhunks/vcr-design.Rmd

```diff
@@ -0,0 +1,152 @@
+## Design
+
+This section explains `vcr`'s internal design and architecture.
+
+## Where vcr comes from and why R6
+
+`vcr` was "ported" from the Ruby gem (aka, library) of the same name[^1]. 
+Because it was ported from Ruby, an object-oriented programming language
+I thought it would be easier to use an object system in R that most
+closely resemble that used in Ruby (at least in my opinion). This 
+thinking lead to choosing [R6][]. The exported functions users interact
+with are not R6 classes, but are rather normal R functions. However,
+most of the internal code in the package uses R6. Thus, familiarity
+with R6 is important for people that may want to contribute to `vcr`,
+but not required at all for `vcr` users.
+
+## Principles
+
+### Mullet design: business up front, party in the back
+
+As described above, `vcr` uses R6 internally, but users interact with
+normal R functions. 
+
+```r
+functions == business
+R6 == party
+```
+
+The business and party could be reversed, depending on your perspective.
+Anyway, the point is that there's two different major sets of code 
+used in `vcr`. 
+
+### Class/function names are inherited from Ruby vcr
+
+Since R `vcr` was ported from Ruby, we kept most of the names of 
+functions/classes and variables. So if you're wondering about why 
+a function, class, or variable has a particular name, its derivation 
+can not be found out in this package, for the most part that is.
+
+### Hooks into http clients
+
+Perhaps the most fundamental thing about that this package work is
+how it knows what HTTP requests are being made. This stumped me for 
+quite a long time. When looking at Ruby vcr, at first I thought it
+must be "listening" for HTTP requests somehow. Then I found out about
+[monkey patching][mp] (see also [this blog post as a good 
+summary][mpblog]); that's how it's achieved in Ruby. That is, the Ruby
+vcr package literally overrides certain methods in Ruby HTTP clients, 
+hijacking internals of the HTTP clients. 
+
+However, monkey patching is not allowed in R. Thus, in R we have to
+somehow have "hooks" into HTTP clients in R. Fortunately, Scott is the
+maintainer of one of the HTTP clients, `crul`, so was able to quickly
+create a hook. Very fortunately, there was already a hook mechanism
+in the `httr` package. 
+
+The actual hooks are not in `vcr`, but in `webmockr`. `vcr` depends on
```

Comment:
a blog post about how the hooks work (that could then get integrated into webmockr docs) would be great!

Reply:
good idea, probably won't have time for that though